### PR TITLE
Make xcm-simulator async with more tests

### PR DESCRIPTION
**Pull request by 4meta5**
Takes work from #3546 and puts them on top of master.

Adds tests from #3537 to xcm-simulator (except teleport_assets)

Also makes message processing asynchronous and pulls it out from the test externalities environment to allow for the processing of messages that go back to the origin chain. The previous implementation ran into borrowing issues in the underlying `RefCell`s when sending messages back to the origin. Messages are processed after the closure passed to `test_ext.execute_with()`.

### CODE REVIEW

**Comment by 4meta5** [comment]
Request review by @shaunxw @kianenigma @apopiak @shawntabrizi and anyone else interested in the xcm-simulator

cc @girazoki for test examples using xcm-simulator

**Review by apopiak — APPROVED** [review]
LGTM

File: xcm/xcm-simulator/example/src/parachain.rs [hunk 1]
```diff
@@ -166,6 +166,12 @@ pub mod mock_msg_queue {
 	#[pallet::getter(fn parachain_id)]
 	pub(super) type ParachainId<T: Config> = StorageValue<_, ParaId, ValueQuery>;
 
+	#[pallet::storage]
+	#[pallet::getter(fn received_dmp)]
+	/// A queue of received DMP messages
+	pub(super) type ReceivedDmp<T: Config> =
+		StorageValue<_, Vec<(MultiLocation, Xcm<T::Call>)>, ValueQuery>;
```
Comment (KiChjang) [inline xcm/xcm-simulator/example/src/parachain.rs:173]: Does this really need a `MultiLocation` when it's always assumed that DMP messages come from the parent, i.e. the relay chain?

File: xcm/xcm-simulator/example/src/parachain.rs [hunk 2]
```diff
@@ -266,7 +272,8 @@ pub mod mock_msg_queue {
 						Self::deposit_event(Event::UnsupportedVersion(id));
 					},
 					Ok(Ok(x)) => {
-						let outcome = T::XcmExecutor::execute_xcm(Parent.into(), x, limit);
+						let outcome = T::XcmExecutor::execute_xcm(Parent.into(), x.clone(), limit);
+						<ReceivedDmp<T>>::mutate(|y| y.push((Parent.into(), x)));
```
Comment (KiChjang) [inline xcm/xcm-simulator/example/src/parachain.rs:276]:
```suggestion
						<ReceivedDmp<T>>::append((Parent.into(), x));
```

But with the `MultiLocation` changes above, you might need this instead:

```suggestion
						<ReceivedDmp<T>>::append(x);
```

File: xcm/xcm-simulator/example/src/parachain.rs (hunk 1 above)
Comment (4meta5) [reply xcm/xcm-simulator/example/src/parachain.rs:173]: Good point, I'll make this change

**Comment by apopiak** [comment]
Let's get this in soon?

**Comment by 4meta5** [comment]
The previous PR which contained the same code (#3546 ) was approved by shaunxw and kianenigma . This PR is approved by @apopiak and was reviewed by keith.

Is there anything blocking this @shawntabrizi ?

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**